src/driver/kvm.rs

```diff
@@ -474,6 +455,884 @@ mod tests {
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock.expect_pause().times(1).returning(|| Ok(()));
+        kvmi_mock
+            .expect_wait_and_pop_event()
+            .times(1)
+            .returning(|_| {
+                Ok(Some(KVMiEvent {
+                    vcpu: 0,
+                    ev_type: KVMiEventType::PauseVCPU,
+                    ffi_event: std::ptr::null_mut(),
+                }))
+            });
+        kvmi_mock.expect_reply().times(1).returning(|_, _| {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "something went wrong",
+            ))
+        });
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+
+        let _result_pause = kvm.pause();
+        let result = kvm.resume();
+
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_read_physical_succeeds() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        let mut buffer: [u8; 4096] = [0; 4096];
+        kvmi_mock
+            .expect_read_physical()
+            .times(1)
+            .returning(|_, buffer| {
+                buffer[0] = 1;
+                Ok(())
+            });
+        let kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+
+        let paddr: u64 = 0;
+        let _result = kvm.read_physical(paddr, &mut buffer);
+
+        assert_eq!(1, buffer[0]);
+    }
+
+    #[test]
+    fn test_read_physical_fails_if_kvmi_read_physical_fails() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock.expect_read_physical().times(1).returning(|_, _| {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "something went wrong",
+            ))
+        });
+        let kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+
+        let paddr: u64 = 0;
+        let mut buffer: [u8; 4096] = [0; 4096];
+        let result = kvm.read_physical(paddr, &mut buffer);
+
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_write_physical_succeeds() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        let paddr: u64 = 0;
+        let mut buffer: [u8; 4096] = [0; 4096];
+        kvmi_mock
+            .expect_write_physical()
+            .times(1)
+            .with(eq(0), function(|x| matches!(x, _buffer)))
```

Comment:
1. You should compare with `paddr`, not with hardcoded `0`
2. `matches!` is only for enum matching. What you are doing here makes no sense and will always return true. You could write `matches!(x, bla)` for the same effect. I encourage you to look at the expanded macro: `cargo expand --features kvm --lib --tests` (You need to install cargo-expand first)